### PR TITLE
Add support for Frozen Notifications

### PR DESCRIPTION
**Pull request by cmelchior**
This PR adds support for frozen notifications and thus adds the final pieces for the move to a fully frozen architecture.

**Design Notes:**
- Open question: What should the initial element of the Flow be? The current element or the first update? Right now, the current element is emitted (which actually might be different from the frozen object), but it is unclear if that is just annoying in the majority of cases? 
- Open question: How to handle a deleted List/Object? A null emission or just close the Flow? Right now we emit `null`, then complete the Flow. This is to force users to handle the situation. Is this too annoying. Should we just complete the flow instead?

**TODO:**
- [x] Injecting Dispatchers through the RealmConfiguration - #307 
- [x] Testing dispatcher configuration - #307
- [x] Test observe on all classes (except List and Realm)
- [ ] Add support for addChangeListener in public API + tests (another PR) 
- [ ] Correctly shutting down the notifier
- [x] Opening and closing the same Realm and creating the same notifier is causing stability isssues.
- [x] findLatest - Pulled out into #306 
- [x] findLatest tests - Pulled out into #306 
- [x] Closing Realm should close all frozen instances ( https://github.com/realm/realm-kotlin/pull/300)
- [x] Refactor NotificationTests to be more readable.
- [x] Changelog

### CODE REVIEW

File: packages/library/src/commonMain/kotlin/io/realm/RealmObject.kt [hunk 1]
```diff
@@ -89,3 +90,31 @@ public fun RealmObject.isValid(): Boolean {
...
+}
+
+public fun <T : RealmObject> T.observe(): Flow<T?> {
+    checkNotificationsAvailable()
+    val internalObject = this as RealmObjectInternal
+    @Suppress("UNCHECKED_CAST")
+    return (internalObject.`$realm$Owner` as RealmReference).owner.observeObject(this as T)
+}
+
+private fun RealmObject.checkNotificationsAvailable() {
```
Comment (rorbech) [inline packages/library/src/commonMain/kotlin/io/realm/RealmObject.kt:108]: Couldn't we split all the notification registration methods into its own interface and implement it only in the supporting classes and then just include a check for `owner is NotifiableRealm`. And then in all the calling methods we could just dispatch it by delegation. Probably for a separate PR that splits Realm into its interface and implementation.

**Review by rorbech — CHANGES_REQUESTED** [review]
Nice, seeing how the pieces are starting to fit together. 

Overall comments from the _first round review_, as I would have to dive a bit more to get the final bits:
- Maybe I am just missing some level of indirection, but seems like all the callback APIs are actually registering on the thread that calls it. 
- Seems like we should prioritize to the the behavioral interface in place soon as we reimplement more and more common stuff instead of abstracting it to `NotifiableRealm`, `RealmLifeCycle`, etc. 

Otherwise just left some comments and skipped commented on all the FIXMEs/TODOs

File: packages/library/src/commonMain/kotlin/io/realm/internal/NotifierRealm.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    val errorMessage = "Flows are not implemented for Notifier Realms, use the callback API instead."
+
+    internal constructor(configuration: RealmConfiguration, dispatcher: CoroutineDispatcher) :
+        super(configuration, RealmInterop.realm_open(configuration.nativeConfig, dispatcher))
+
+    override fun <T : RealmObject> addResultsChangeListener(
+        results: RealmResults<T>,
+        callback: Callback<RealmResults<T>>
+    ): Cancellable {
+        throw UnsupportedOperationException(errorMessage)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/NotifierRealm.kt:27]: Sounds fine with me 👍

File: packages/library/src/commonMain/kotlin/io/realm/internal/NotificationToken.kt [hunk 3]
```diff
@@ -19,16 +19,18 @@ package io.realm.internal
...
+import kotlinx.atomicfu.atomic
 
 class NotificationToken<T>(t: T, private val token: NativePointer) : Cancellable {
 
-    private var t: T? = t
+    private var t: AtomicRef<T?> = atomic(t)
 
     override fun cancel() {
-        if (t != null) {
+        if (t.value != null) {
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/NotificationToken.kt:30]: No, releasing the token multiple times actually crashes, so we should probably guard this

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+/**
+ * Class responsible for controlling notifications for a Realm. It does this by wrapping a live Realm on which
+ * notifications can be registered. Since all objects that is otherwise exposed to users are frozen, they need
+ * to be thawed when reaching the live Realm.
+ *
+ * For Lists and Objects, this can result in the object no longer existing. In this case, Flows complete
+ * immediately and changelisteners will never be triggered (Is this the semantics we want? What do Room do?).
+ *
+ * Notifications are not allowed on live objects. But it is assumed that other layers check that invariant before
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:34]: Generally, we haven't used this kind of Asserts before, so I would get a bit weary adding it in this PR, but probably a topic we should discuss in general

File: test/src/androidTest/kotlin/io/realm/InstrumentedTests.kt [hunk 5]
```diff
@@ -40,7 +41,7 @@ class InstrumentedTests {
     @Before
     fun setup() {
         tmpDir = PlatformUtils.createTempDir()
-        val configuration = RealmConfiguration(path = "$tmpDir/default.realm", schema = setOf(Sample::class))
+        val configuration = RealmConfiguration(path = "$tmpDir/${createRandomString(16)}.realm", schema = setOf(Sample::class))
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/InstrumentedTests.kt:44]: Yes, I might have run into other issues that somehow assumed this would fix it, but tmpDir should make unique paths for each test already.

File: test/src/commonMain/kotlin/io/realm/util/Utils.kt [hunk 6]
```diff
@@ -32,13 +36,28 @@ expect object PlatformUtils {
...
 }
+
+/**
+ * Helper method for easily updating a single object. The updated object will be returned.
+ * This method control its own write transaction, so cannot be called inside a write transaction
+ */
+suspend fun <T : RealmObject> T.update(block: T.() -> Unit): T {
+    val realm = ((this as RealmObjectInternal).`$realm$Owner` as RealmReference).owner as Realm
+    return realm.write {
+        val liveObject: T = findLatest(this@update)!!
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/commonMain/kotlin/io/realm/util/Utils.kt:59]: This was mostly for testing where it would be a mistake if the object was deleted. But I'll move the method to the public API and change it a little bit.

File: test/src/commonMain/kotlin/io/realm/util/Utils.kt [hunk 7]
```diff
@@ -32,13 +36,28 @@ expect object PlatformUtils {
...
     fun printlntid(message: String) {
         println("<" + PlatformUtils.threadId() + "> $message")
     }
 }
+
+/**
+ * Helper method for easily updating a single object. The updated object will be returned.
+ * This method control its own write transaction, so cannot be called inside a write transaction
+ */
+suspend fun <T : RealmObject> T.update(block: T.() -> Unit): T {
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/commonMain/kotlin/io/realm/util/Utils.kt:56]: If we did that, isn't it basically equivalent to `findLatest`? 🤔

File: packages/library/src/commonMain/kotlin/io/realm/RealmObject.kt (hunk 1 above)
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/RealmObject.kt:108]: Yes, sounds like something we should think about, but will probably be overkill for this PR

File: packages/cinterop/src/darwinCommon/kotlin/io/realm/interop/RealmInterop.kt [hunk 8]
```diff
@@ -651,8 +651,9 @@ actual object RealmInterop {
                 // Use the callback as user data
                 StableRef.create(callback).asCPointer(),
                 staticCFunction<COpaquePointer?, Unit> { userdata ->
-                    userdata?.asStableRef<Callback>()?.dispose()
-                        ?: error("Notification callback data should never be null")
+                    // FIXME This throws Uncaught Kotlin exception: kotlin.native.IncorrectDereferenceException: illegal attempt to access non-shared <object>@a742f038 from other thread
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/cinterop/src/darwinCommon/kotlin/io/realm/interop/RealmInterop.kt]: Freezing our notification objects seems to fix the problem

File: test/src/commonMain/kotlin/io/realm/util/Utils.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (rorbech) [reply test/src/commonMain/kotlin/io/realm/util/Utils.kt:56]: Yeah, I guess so. I had some memory of a more tedious pattern. If we can get away with a one-liner `findLatest` I guess there is no need for an unwrapped write version, and might be that this wrapped write version is convenient to have around.

File: test/src/commonMain/kotlin/io/realm/util/Utils.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (rorbech) [reply test/src/commonMain/kotlin/io/realm/util/Utils.kt:59]: Ok. I guess we can discuss if we consider to add a similar method in the actual library.

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            throw IllegalStateException("Notification could not be sent: $result")
+        }
+    }
+
+    // FIXME Need to expose change details to the user
+    //  https://github.com/realm/realm-kotlin/issues/115
+    /**
+     * Register a change listener on a live RealmResults. All objects returned in the callback are frozen.
+     */
+    fun <T : RealmObject> addResultsChangedListener(results: RealmResults<T>, callback: Callback<RealmResults<T>>): Cancellable {
```
Comment (rorbech) [inline packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:128]: This must be called on the notifier thread? Otherwise we wouldn't be able to thaw the result and actually register for notifications!? Maybe add a comment and make it internal.

**Review by rorbech — APPROVED** [review]
I guess there are some rough edges around closing and threading for callback APIs, but as we are targeting the feature branch, I guess it is probably OK and can be targeted in another PR. 

Maybe just a matter of naming, but the whole chain of calls for registering for notifications is a bit hard to follow. Maybe make the "inner" methods of the `SuspendableNotifier` private and name them something to indicate that they are internal. 
Further, as far as I can see, there is still a single method for registering for notifications through a callback API that allows registering for notification on the wrong thread.

**Comment by rorbech** [comment]
> Further, as far as I can see, there is still a single method for registering for notifications through a callback API that allows registering for notification on the wrong thread.

Seems like the callback APIs are in fact internal (just only marked in the interface and not in `Realm`), so should be good enough for now.

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    fun tearDown() {
+        if (!realm.isClosed()) {
+            realm.close()
+        }
+        PlatformUtils.deleteTempDir(tmpDir)
+    }
+
+    @Test
+    override fun observe() {
+        runBlocking {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:48]: Hmm. I guess it was to avoid a lot of `1. Method ... should be void` 🤔

File: packages/library/src/commonMain/kotlin/io/realm/BaseRealm.kt [hunk 11]
```diff
@@ -33,6 +34,10 @@ public abstract class BaseRealm internal constructor(
     dbPointer: NativePointer
 ) {
 
+    companion object {
+        const val flowNotSupportMessage = "Flows are not implemented for this type of Realm, use the callback API instead."
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/BaseRealm.kt:38]: True, I'll reword to something better. I'm overriding this error message in places that might leak to the public anyway.

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+     * This flow is guaranteed to emit before any other streams listening to individual objects or
+     * query results.
+     */
+    fun realmChanged(): Flow<Pair<NativePointer, VersionId>> {
+        // FIXME Workaround until proper Realm Changed Listeners are implemented
+        return _realmChanged.asSharedFlow()
+            .map {
+                Pair(it.dbPointer, VersionId(RealmInterop.realm_get_version_id(it.dbPointer)))
+            }.distinctUntilChanged { old, new ->
+                old.second == new.second
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:66]: I also need to remove this again. It is causing memory leaks if the element is dropped as Realm doesn't know it need to clean it up.

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt [hunk 13]
```diff
@@ -16,21 +16,34 @@
...
+import kotlinx.coroutines.launch
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 
 // TODO API-PUBLIC Document platform specific internals (RealmInitilizer, etc.)
 class Realm private constructor(configuration: RealmConfiguration, dbPointer: NativePointer) :
     BaseRealm(configuration, dbPointer) {
 
-    private val writer: SuspendableWriter = SuspendableWriter(this)
+    internal val realmScope: CoroutineScope = CoroutineScope(SupervisorJob() + configuration.notificationDispatcher)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt:42]: True, this came from when I thought user actions would somehow also be using this. Although I suspect we will use this scope internally for the public callback API, so maybe better to just keep it?

File: packages/library/src/commonMain/kotlin/io/realm/internal/NotifierRealm.kt [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+import io.realm.BaseRealm
+import io.realm.RealmConfiguration
+import io.realm.interop.RealmInterop
+import kotlinx.coroutines.CoroutineDispatcher
+
+/**
+ * Live Realm on which notifications can be registered.
+ */
+internal class NotifierRealm : BaseRealm {
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/NotifierRealm.kt:11]: BaseRealm is currently public because it holds properties available to the user (so we cannot make it internal). Keeping it abstract at least makes sure that people don't accidentally create instances of it.

[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/NotifierRealm.kt:11]: But I agree, this class be removed

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    fun addRealmChangedListener(callback: Callback<Pair<NativePointer, VersionId>>): Cancellable {
+        TODO("Waiting for RealmInterop to have support for global Realm changed")
+    }
+
+    /**
+     * Listen to changes to a RealmResults
+     */
+    fun <T : RealmObject> resultsChanged(results: RealmResults<T>): Flow<RealmResults<T>> {
+        return callbackFlow {
+            val token: AtomicRef<Cancellable> = kotlinx.atomicfu.atomic(NO_OP_NOTIFICATION_TOKEN)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:94]: We are using the NO_OP_NOTIFICATION_TOKEN in a few places as it makes the control flow a little nicer, e.g. when registering a Object change listener. So I just used the same pattern everywhere in this class.

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+
+    // FIXME Need to expose change details to the user
+    //  https://github.com/realm/realm-kotlin/issues/115
+    /**
+     * Register a change listener on a live RealmResults. All objects returned in the callback are frozen.
+     */
+    fun <T : RealmObject> addResultsChangedListener(results: RealmResults<T>, callback: Callback<RealmResults<T>>): Cancellable {
+        val liveResults = results.thaw(realm.realmReference)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:129]: No, it is a query result, so the only times this becomes invalid if is the table is completely removed (which isn't possible) or if the Realm is closed (which is caught earlier)

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+            awaitClose {
+                token.value.cancel()
+            }
+        }
+    }
+
+    // FIXME Need to expose change details to the user
+    //  https://github.com/realm/realm-kotlin/issues/115
+    /**
+     * Listen to changes to a RealmObject through a change listener. The callback will happen
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:184]: Yes, I added a comment about this.

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:48]: Yes, not exactly sure why it happens when overriding methods but not when doing directly. I suspect we can fix this if we introduce our own `runBlockingTest` that returns `Unit`.

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmResultsNotificationsTests.kt [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+            val c = Channel<Int>(capacity = 1)
+            val counter = AtomicInt(0)
+            val observer1 = async {
+                realm.objects(Sample::class).observe().collect {
+                    when (counter.incrementAndGet()) {
+                        1 -> c.trySend(it.size)
+                        2 -> {
+                            realm.close()
+                            c.trySend(-1)
+                            println("realm closed")
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmResultsNotificationsTests.kt:147]: Yes, this test isn't done yet. I marked it with a comment.

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmResultsNotificationsTests.kt [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+                        }
+                    }
+                }
+            }
+            val observer2 = async {
+                realm.objects(Sample::class).observe().collect {
+                    println(it.first().stringField)
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmResultsNotificationsTests.kt:154]: Yes, this test isn't done yet. I marked it with a comment.

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt [hunk 20]
```diff
@@ -195,6 +265,8 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
         runBlocking {
             realmPointerMutex.withLock {
                 writer.close()
+                notifierJob.value?.cancel()
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt:268]: No 🤦

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+                    c.trySend(it)
+                }
+                // Emit sentinel value to signal that flow completed
+                c.send(Sample())
+            }
+            assertNotNull(c.receive())
+            realm.write {
+                delete(findLatest(obj)!!)
+            }
+            assertNull(c.receive()) // Null is sent when object is deleted
```
Comment (nhachicha) [inline test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:145]: Shouldn't we close/complete the Flow if the element we're observing is deleted?

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+     * on the configured [SuspendableNotifier.dispatcher] thread.
+     */
+    fun <T : RealmObject> addListChangedListener(list: List<T>, callback: Callback<List<T>>): Cancellable {
+        TODO("Implement and convert method to use RealmList when available")
+    }
+
+    fun close() {
+        // FIXME Is it safe at all times to close a Realm? Probably not during a changelistener callback, but Mutexes
+        //  are not supported within change listeners as they are not suspendable.
+        runBlocking(dispatcher) {
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:232]: Yes, I still haven't convinced myself either way yet. But so far all simple unit tests seem to pass. I'm making a TODO list for this so we remember to come back to it.

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt (hunk 9 above)
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:128]: Yes, I added a comment in the Dokka about it. So far this is only being used by our Flows

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    @Test
+    @Ignore
+    override fun closeRealmInsideFlowThrows() {
+        TODO("Not yet implemented")
+    }
+
+    @Test
+    override fun closingRealmDoesNotCancelFlows() {
```
Comment (nhachicha) [inline test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:159]: This leads to undefined behaviour in Core, since we still emit managed objects accessing a closed Realm!

File: test/build.gradle.kts [hunk 24]
```diff
@@ -54,6 +54,10 @@ kotlin {
 
         getByName("commonTest") {
             dependencies {
+                // TODO AtomicFu doesn't work on the test project. It throws:
+                //  Source set 'androidMain' of project 'test' is part of several compilations [debug, release]
```
Comment (nhachicha) [inline test/build.gradle.kts:58]: https://github.com/Kotlin/kotlinx.atomicfu/issues/90#issuecomment-597872907 :(

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+
+/**
+ * Class responsible for controlling notifications for a Realm. It does this by wrapping a live Realm on which
+ * notifications can be registered. Since all objects that is otherwise exposed to users are frozen, they need
```
Comment (nhachicha) [inline packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:28]:
```suggestion
 * notifications can be registered. Since all objects that are otherwise exposed to users are frozen, they need
```

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+ * to be thawed when reaching the live Realm.
+ *
+ * For Lists and Objects, this can result in the object no longer existing. In this case, Flows complete
+ * immediately and changelisteners will never be triggered (Is this the semantics we want? What do Room do?).
+ *
+ * Notifications are not allowed on live objects. But it is assumed that other layers check that invariant before
+ * methods on this class is called.
```
Comment (nhachicha) [inline packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt]:
```suggestion
 * methods on this class are called.
```

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (nhachicha) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt:42]: Why one job? can we use this supervisor to also cancel all observers when a Realm is closed?

File: examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt [hunk 27]
```diff
@@ -52,17 +53,11 @@ class MainActivity : AppCompatActivity() {
 
         numATV.addTextChangedListener(textWatcher)
         numBTV.addTextChangedListener(textWatcher)
-    }
 
-    override fun onResume() {
-        super.onResume()
-        registration = Calculator.listen {
-            countTV.text = "History count: ${Calculator.history().size}"
+        lifecycleScope.launch {
```
Comment (nhachicha) [inline examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt:57]: I think there's a value in keeping it inside the `onResume` callback to demonstrate the cancel capability?

**Review by nhachicha — COMMENTED** [review]
Nice to see this taking shape 🎉
I breezed through it quickly and left some minor questions/comments

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt (hunk 23 above)
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:159]: The test is mostly here so we capture whatever behavior we want. Right now it just stops emitting events (because change listeners stop), but the flows themselves are not completed.

We need to figure out exactly what should happen here. For writers we throw an exception, but I'm not sure that is possible for Flows since the even passed a few event queues along the way.

File: test/build.gradle.kts (hunk 24 above)
Comment (cmelchior) [reply test/build.gradle.kts:58]: Bummer. I'll add a reference to it in the comment

File: packages/library/src/commonMain/kotlin/io/realm/Realm.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/Realm.kt:42]: It is because Flows run inside user-defined scopes. I originally thought we might be able to interleave scopes, but that isn't possible.

But I suspect we will end up piping the change listener API through here. For which we don't want everything canceled if one fail.

File: test/src/androidTest/kotlin/io/realm/MemoryTests.kt [hunk 28]
```diff
@@ -165,7 +169,7 @@ class MemoryTests {
     }
 
     private fun openRealmFromTmpDir(): Realm {
-        val configuration = RealmConfiguration(path = "$tmpDir/default.realm", schema = setOf(Sample::class))
+        val configuration = RealmConfiguration(path = "$tmpDir/${createRandomString(16)}.realm", schema = setOf(Sample::class))
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/MemoryTests.kt:172]: It was. I made a brain fart. This change has been reverted.

File: test/src/androidTest/kotlin/io/realm/MemoryTests.kt [hunk 29]
```diff
@@ -50,6 +52,7 @@ class MemoryTests {
     }
 
     @Test
+    @Ignore // Probably related to the frozen Realms in the Notifier not being released as they are dropped inside the Flow to the public Realm
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/MemoryTests.kt:55]: This still fails on Android, but not iOS. Not 100% sure what is causing it yet.

File: test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt (hunk 21 above)
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/shared/notifications/RealmObjectNotificationsTests.kt:145]: Yes, we are also doing that. But there is an open question if we should just complete the Flow directly or emit `null` before doing it, to make sure that people actually handle the case.

Otherwise they need to remember to do something like:

```
scope.launch {
                obj.observe()
                    .onCompletion { 
                        // Object was deleted
                    }
                    .collect {
                        // Handle object update
                }
}
```

Right now I opted for emitting `null`, but that is something we need to agree on. Do you have any preference?

File: examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt (hunk 27 above)
Comment (cmelchior) [reply examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt:57]: But `lifecycleScope` automatically handles the cancelation?

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:218]: But I don't see the comment ;)

File: test/src/commonTest/kotlin/io/realm/NotificationTests.kt [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    @Test
+    fun addChangeListener_emitOnProvidedDispatcher() {
+        // FIXME Implement in another PR
+    }
+
+    @Test
+    fun openSameRealmFileWithDifferentDispatchers() {
+        // FIXME This seems to not work
```
Comment (rorbech) [inline test/src/commonTest/kotlin/io/realm/NotificationTests.kt:53]: Is there still a problem after merging the fix that closes the realm on the correct thread in the SuspendableWriter?

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+        }
+        if (!result.isSuccess) {
+            // TODO Is there a better way to handle this?
+            throw IllegalStateException("Notification could not be sent: $result")
+        }
+    }
+
+    private fun notifyRealmChanged(frozenRealm: RealmReference) {
+        if (!_realmChanged.tryEmit(frozenRealm)) {
+            println("Failed to send update to Realm from the Notifier: ${owner.configuration.path}")
```
Comment (rorbech) [inline packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:158]: WIP?

**Review by rorbech — APPROVED** [review]
Getting there. Still quite a lot of FIXMEs, but I wouldn't mind merging this if we just track them elsewhere.

File: test/src/androidTest/kotlin/io/realm/MemoryTests.kt (hunk 29 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/MemoryTests.kt]: I made a mistake in the file name. THis is working now

[the remark this answers is not in the document]
Comment (cmelchior) [reply test/src/androidTest/kotlin/io/realm/MemoryTests.kt]: It was failing as I had accidentially changed the Realm file name. It works now.

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:218]: Ups, was looking the wrong place. Added

**Review by rorbech — APPROVED** [review]
Great explanatory CHANGELOG. 

LGTM, but remember to add the spill over (close considerations, semantics on initial/deleted elements for flows, etc.) in #291

File: examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt (hunk 27 above)
Comment (nhachicha) [reply examples/kmm-sample/androidApp/src/main/java/io/realm/example/kmmsample/androidApp/MainActivity.kt:57]: After all it's also beneficial to demonstrate both approaches, so let's keep the `lifecycleScope` 👍

File: CHANGELOG.md [hunk 32]
```diff
@@ -1,9 +1,33 @@
 ## 0.3.0 (YYYY-MM-DD)
 
+This release contains a big depature in the architectural design of how Realm is currently implemented. At a high level it moves from "Thread-confined, Live Objects" to "Frozen Objects". The reasons for this shift are discussed [here](https://docs.google.com/document/d/1bGfjbKLD6DSBpTiVwyorSBcMqkUQWedAmmS_VAhL8QU/edit#heading=h.fzlh39twuifc). 
+
+At a high level this has a number of implications:
+
+    1. Only one Realm instance is needed across the entire application.
```
Comment (nhachicha) [inline CHANGELOG.md:7]:
```suggestion
    1. Only one Realm instance (per `RealmConfiguration`) is needed across the entire application. 
```
